tools/pythonpkg/tests/fast/test_scalar_udf.py

```diff
@@ -0,0 +1,205 @@
+
+    def test_detected_parameters(self):
+        def concatenate(a: str, b: str):
+            return a + b
+
+        con = duckdb.connect()
+        con.register_scalar_udf('py_concatenate', concatenate, None, VARCHAR)
+        res = con.sql("""
+            select py_concatenate('5','3');
+        """).fetchall()
+        assert res[0][0] == '53'
+
+    def test_detected_return_type(self):
+        def add_nums(*args) -> int:
+            sum = 0;
+            for arg in args:
+                sum += arg
+            return sum
+
+        con = duckdb.connect()
+        con.register_scalar_udf('add_nums', add_nums)
+        res = con.sql("""
+            select add_nums(5,3,2,1);
+        """).fetchall()
+        assert res[0][0] == 11
+
+    def test_varargs(self):
+        def variable_args(*args):
+            amount = len(args)
+            return amount
+
+        con = duckdb.connect()
+        con.register_scalar_udf('varargs', variable_args, None, BIGINT)
+        res = con.sql("""select varargs('5', '3', '2', 1, 0.12345)""").fetchall()
+        assert res == [(5,)]
+
+    def test_overwrite_name(self):
+        def func(x):
+            return x
+        con = duckdb.connect()
+        # create first version of the function
+        con.register_scalar_udf('func', func, [BIGINT], BIGINT)
+
+        # create relation that uses the function
+        rel1 = con.sql('select func(3)')
+
+        def other_func(x):
+            return x
+
+        with pytest.raises(duckdb.NotImplementedException, match="A function by the name of 'func' is already registered, registering multiple functions by the same name is not supported yet, please unregister it first"):
+            con.register_scalar_udf('func', other_func, [VARCHAR], VARCHAR)
+
+        con.unregister_udf('func')
+
+        with pytest.raises(duckdb.InvalidInputException, match='Catalog Error: Scalar Function with name func does not exist!'):
+            # Attempted to execute the relation using the 'func' function, but it was deleted
+            rel1.fetchall()
+
+        con.register_scalar_udf('func', other_func, [VARCHAR], VARCHAR)
+        # create relation that uses the new version
+        rel2 = con.sql("select func('test')")
+
+        # execute both relations
+        res1 = rel1.fetchall()
+        res2 = rel2.fetchall()
+        # This has been converted to string, because the previous version of the function no longer exists
+        assert res1 == [('3',)]
+        assert res2 == [('test',)]
+
+    def test_nulls(self):
+        def five_if_null(x):
+            if (x == None):
+                return 5
+            return x
+        con = duckdb.connect()
+        con.register_scalar_udf('null_test', five_if_null, [BIGINT], BIGINT, null_handling = "SPECIAL")
+        res = con.sql('select null_test(NULL)').fetchall()
+        assert res == [(5,)]
+
+    def test_exceptions(self):
+        def throws(x):
```

Comment:
Can we add more tests:

* What if we return the wrong typed object (e.g. return type is specified as BIGINT but we return a VARCHAR)
* Could we add more functions with nested types (e.g. returning a list or a map/dictonary, or tuples)?
* Could we add a function that returns `None`?
* Could we add a function that returns `NaN` (float)?
* Could we add a functions that return integers that do not fit (e.g. returntype smallint -> returns `2**30`, returntype bigint, returns `2**70`)?
* Could we add a test with non-inlined generated strings?
* What happens if we call `unregister_udf` with a non-existent UDF, or call it multiple times?
* Can we use registered functions also in relations, e.g. `.project('my_function(x)')`?

Reply:
- added `test_return_incorrectly_typed_object`
- added the "native" version to `test_type_coverage` in `test_pyarrow_udf.py`, extended it for struct + list, added a separate extra test for MAP (`test_map_coverage`) as this doesn't seem to work with prepared parameters, it gets detected as STRUCT instead.
- added `test_return_null`
- added `test_nan`
- added `test_return_overflow` (this only covers `native` though)
- I think this should be covered by `test_type_coverage`

I still need to cover the last two points

Reply:
I've added an extra test to `test_type_coverage` using the `project` method, and I've added the missing tests for `remove_function` to `test_remove_function.py`